torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1276,13 +1302,33 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 #endif
 }
 
-void dump_debugging_info() {
+void ProcessGroupNCCL::registerDebugInfoCallbackStorer(
+    std::unique_ptr<ProcessGroupNCCL::DebugInfoCallbackStorer> callbackStorer) {
+  TORCH_CHECK_WITH(
+      DistBackendError,
+      debugInfoCallbackStorer_ == nullptr,
+      "ProcessGroupNCCL DebugInfoCallbackStorer already registered");
+  debugInfoCallbackStorer_ = std::move(callbackStorer);
+}
+
+void ProcessGroupNCCL::dumpDebuggingInfo() {
   LOG(ERROR)
       << "No PGNCCL's watchdog heartbeat detected, so we are dumping debug info.";
   if (parseEnvVarIntDefault("TORCH_NCCL_TRACE_BUFFER_SIZE", 0) > 0) {
-    // TODO: Find the right and proper way to dump the debug info.
-    // We cannot print out debugging info directly.
-    LOG(ERROR) << "nccl_trace: " << dump_nccl_trace();
+    // We dump nccl trace into local disk by default and users can register
+    // their own storer via `registerDebugInfoCallbackStorer`.
+    auto ncclTrace = dump_nccl_trace();
+    if (debugInfoCallbackStorer_ == nullptr) {
+      // Dump the trace blob into local disk as a fallback.
+      const char* fileName = parseEnvVarString(
```

Comment:
its a little odd to me to "just in time" parse the env and construct the default writer.  i wonder why not create it inside the constructor?  Do we benefit from having ability to change the writer mid-way through the program?

Reply:
Actually I don't have a strong opinion on this. To be passing in the constructor vs getting the path name in the constructor is pretty much the same.